core/metrics-core-service/src/main/java/org/hawkular/metrics/core/service/DataAccessImpl.java

```diff
@@ -223,17 +341,185 @@
+        int key = ordinal;
+        key |= code << 24;
+        return key;
+    }
+
+    private Integer getMapKey(MetricType type, TempStatement ts) {
+        return getMapKey(type.getCode(), ts.ordinal());
+    }
+
+    void prepareTempStatements(String tableName, Long mapKey) {
+        // Create an entry to the correct Long
+        Map<Integer, PreparedStatement> statementMap = new HashMap<>();
+
+        // Per metricType
+        for (MetricType<?> metricType : MetricType.userTypes()) {
+            if(metricType == STRING) { continue; } // We don't support String metrics in temp tables yet
+            for (TempStatement st : TempStatement.values()) {
+                Integer key = getMapKey(metricType, st);
+
+                String formatSt;
+                switch(st.getType()) {
+                    case READ:
+                        formatSt = String.format(st.getStatement(), metricTypeToColumnName(metricType),
+                                tableName);
+                        break;
+                    case WRITE:
+                        formatSt = String.format(st.getStatement(), tableName,
+                                metricTypeToColumnName(metricType));
+                        break;
+                    default:
+                        // Not supported
+                        continue;
+                }
+
+                PreparedStatement prepared = session.prepare(formatSt);
+                statementMap.put(key, prepared);
+            }
+        }
+        // Untyped
+        for (TempStatement st : TempStatement.values()) {
+            Integer key = getMapKey(MetricType.UNDEFINED, st);
+            String formatSt;
+            switch(st.getType()) {
+                case SCAN:
+                case CREATE:
+                case DELETE:
+                    formatSt = String.format(st.getStatement(), tableName);
+                    break;
+                default:
+                    continue;
+            }
+            PreparedStatement prepared = session.prepare(formatSt);
+            statementMap.put(key, prepared);
+        }
+        prepMap.put(mapKey, statementMap);
+    }
+
+    @Override
+    public Observable<ResultSet> createTempTablesIfNotExists(final Set<Long> timestamps) {
+        return Observable.fromCallable(() -> {
+            Set<String> tables = timestamps.stream()
+                    .map(this::getTempTableName)
+                    .collect(Collectors.toSet());
+
+            // TODO This is an IO operation..
+            metadata.getKeyspace(session.getLoggedKeyspace()).getTables().stream()
+                    .map(AbstractTableMetadata::getName)
+                    .filter(t -> t.startsWith(TEMP_TABLE_NAME_PROTOTYPE))
+                    .forEach(tables::remove);
+
+            return tables;
+        })
+
+                .flatMapIterable(s -> s)
+                .zipWith(Observable.interval(300, TimeUnit.MILLISECONDS), (st, l) -> st)
```

Comment:
300 ms is still completely arbitrary. The slowest and safest way is probably to use ```Metadata.checkSchemaAgreement```.